gson/src/main/java/com/google/gson/internal/bind/TypeAdapters.java

```diff
@@ -368,7 +374,10 @@ public Number read(JsonReader in) throws IOException {
         public void write(JsonWriter out, Number value) throws IOException {
           if (value == null) {
             out.nullValue();
+          } else if (value instanceof Long) {
+            out.value(value);
           } else {
+            // Only perform unwrapping if numeric conversion is necessary
             out.value(value.longValue());
```

Comment:
For correctness this `value instanceof Long` check is not needed; the previous code which always called `out.value(value.longValue())` was fine as well. However, for `JsonTreeWriter` it would have unwrapped and then wrapped the value again. But on the other hand for `JsonWriter` `value(long)` might be more efficient. I am not really sure which approach is better for performance, or if it even matters at all.

(same applies to `Double` below)